caffe2/operators/onnxifi_op.cc

```diff
@@ -0,0 +1,192 @@
+#include "caffe2/operators/onnxifi_op.h"
+
+namespace caffe2 {
+
+namespace {
+
+void CPUTensorToTensorProto(
+    const TensorCPU& cpu_tensor,
+    ::ONNX_NAMESPACE::TensorProto* t) {
+  const auto len = cpu_tensor.size();
+  if (cpu_tensor.template IsType<float>()) {
+    t->set_data_type(::ONNX_NAMESPACE::TensorProto::FLOAT);
+    const float* data = cpu_tensor.template data<float>();
+    for (auto i = 0; i < len; ++i) {
+      t->add_float_data(*data++);
+    }
+  } else if (cpu_tensor.template IsType<int64_t>()) {
+    t->set_data_type(::ONNX_NAMESPACE::TensorProto::INT64);
+    const int64_t* data = cpu_tensor.template data<int64_t>();
+    for (auto i = 0; i < len; ++i) {
+      t->add_int64_data(*data++);
+    }
+  } else if (cpu_tensor.template IsType<int32_t>()) {
+    t->set_data_type(::ONNX_NAMESPACE::TensorProto::INT32);
+    const int32_t* data = cpu_tensor.template data<int32_t>();
+    for (auto i = 0; i < len; ++i) {
+      t->add_int32_data(*data++);
+    }
+  } else {
+    CAFFE_THROW(
+        "Don't know how to convert workspace tensor type ",
+        cpu_tensor.meta().name(),
+        " to ONNX TensorProto");
+  }
+}
+
+void BlobToTensorDescriptor(
+    const std::string& name,
+    Workspace* ws,
+    onnxTensorDescriptor* desc,
+    std::vector<std::vector<uint64_t>>* shapes) {
+  const Blob* blob = ws->GetBlob(name);
+  CAFFE_ENFORCE(blob, "Blob ", name, " doesn't exist");
+
+  // Memory type
+  // We only allow weights to be CPU tensor for now
+  CAFFE_ENFORCE(
+      blob->template IsType<TensorCPU>(),
+      "Initialization blob ",
+      name,
+      " needs to be TensorCPU");
+  desc->memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+
+  // Data type
+  const auto& cpu_tensor = blob->template Get<TensorCPU>();
+  if (cpu_tensor.template IsType<float>()) {
+    desc->dataType = ONNXIFI_DATATYPE_FLOAT32;
+    desc->buffer = (onnxPointer)(cpu_tensor.data<float>());
```

Comment:
`reinterpret_cast<onnxPointer>(cpu_tensor.data<float>());` would be more C++-style